filebeat/input/v2/statestore/store.go

```diff
@@ -0,0 +1,176 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package statestore
+
+import (
+	"sync"
+
+	"github.com/elastic/go-concert"
+	"github.com/elastic/go-concert/atomic"
+
+	"github.com/elastic/beats/libbeat/registry"
+)
+
+// Store provides some coordinates access to a registry.Store.
+// All update and read operations require users to acquire a resource first.
+// A Resource must be locked before it can be modified. This ensures that at most
+// one go-routine has access to a resource. Lock/TryLock/Unlock can be used to
+// coordinate resource access even between independent components.
+type Store struct {
+	active atomic.Bool
+
+	session *storeSession
+	*store
+}
+
+// store is the shared store instance as is tracked by the Connector.
+// Any two go-routines accessing a Store will reference the same underlying store.
+// If one of the go-routines closes the store, we ensure that the shared resources
+// are kept alive until all go-routines have given up access.
+type store struct {
```

Comment:
Could you please rename this `struct`? It is a bit confusing to have the same name for both an exported and an unexported type. My suggestion would be `sharedStore`.